factfinder/main.py

```diff
@@ -120,74 +224,107 @@ def create_variable(self, pff_variable) -> Variable:
-                )
-        elif geotype == "cd_fp_100":
-            if v.source == "decennial":
-                df = self.aggregate_horizontal(client, v, "block")
-                df = self.aggregate_vertical(
-                    df, from_geotype="block", to_geotype="cd_fp_100"
-                )
-            else:
-                df = self.aggregate_horizontal(client, v, "block group")
-                df = self.aggregate_vertical(
-                    df, from_geotype="block group", to_geotype="cd_fp_100"
+        from_geotype, aggregate_vertical = self.get_aggregate_vertical(
+            v.source, geotype
+        )
+        df = self.aggregate_horizontal(client, v, from_geotype)
+        df = aggregate_vertical(df)
+        return df
+
+    def calculate_c_e_m_p_z(self, v:Variable, geotype:str) -> pd.DataFrame:
+        """
+        this function will calculate e, m first, then based on if the 
+        variable is a median variable or base variable, it would calculate
+        p and z accordingly. Note that c calculation is the same for all variables
+        """
+        output_fields = [
+            "census_geoid",
+            "pff_variable",
+            "geotype",
+            "c",
+            "e",
+            "m",
+            "p",
+            "z",
+        ]
+        # If pff_variable is a median variable, then we would need
+        # to calculate using calculate_median_e_m for aggregated geography
+        # there's no need to calculate p, z for median variables
+        if v.pff_variable in self.median_variables:
+            df = (
+                self.calculate_median_e_m(v.pff_variable, geotype)
+                if geotype in self.aggregated_geography
+                else self.calculate_e_m(v.pff_variable, geotype)
+            )
+            df["p"] = 100 if geotype in ["city", "borough"] else np.nan
+            df["z"] = np.nan
+        else:
+            df = self.calculate_e_m(v.pff_variable, geotype)
+            # If pff_variable is not base_variable, then p,z
+            # are calculated against the base variable e(agg_e), m(agg_m)
+            if v.pff_variable not in self.base_variables:
+                df = self.calculate_e_m(v.pff_variable, geotype)
+                df_base = self.calculate_e_m(v.base_variable, geotype)
+                df = df.merge(
+                    df_base[["census_geoid", "e", "m"]].rename(
+                        columns={"e": "agg_e", "m": "agg_m"}
+                    ),
+                    how="left",
+                    on="census_geoid",
                 )
-        elif geotype == "cd_park_access":
-            if v.source == "decennial":
-                df = self.aggregate_horizontal(client, v, "block")
-                df = self.aggregate_vertical(
-                    df, from_geotype="block", to_geotype="cd_park_access"
+                del df_base
+                df["p"] = df.apply(lambda row: get_p(row["e"], row["agg_e"]), axis=1)
```

Comment:
Will we always calculate p and z by default, rather than trying to download them using _PE and _PM census variables? This is more straightforward. Are there any disadvantages to calculating ourselves each time?

Reply:
oh right, forgot about the PE PM issue, this would need to be addressed
calculated PE PM are usually off compared to census PE PM, population prefers to take census PE PM whenever available